ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/preexecution/PreExecutingValidatingCommitter.scala

```diff
@@ -0,0 +1,94 @@
+  * alongside values), parametric in the logic that produces a fingerprint given a value.
+  *
+  * @param now The record time provider.
+  * @param keySerializationStrategy The key serializer used for state keys.
+  * @param validator The pre-execution validator.
+  * @param valueToFingerprint The logic that produces a fingerprint given a value.
+  * @param postExecutionFinalizer The post-execution finalizer that will also perform conflicts detection and
+  *                               time bounds checks.
+  * @param stateValueCache The cache instance for state values.
+  * @param cacheUpdatePolicy The caching policy for values.
+  * @param materializer The Akka materializer.
+  * @tparam LogResult type of the offset used for a log entry.
+  */
+class PreExecutingValidatingCommitter[LogResult](
+    now: () => Instant,
+    keySerializationStrategy: StateKeySerializationStrategy,
+    validator: PreExecutingSubmissionValidator[AnnotatedRawKeyValuePairs],
+    valueToFingerprint: Option[Value] => Fingerprint,
+    postExecutionFinalizer: PostExecutionFinalizerWithFingerprintsFromValues[LogResult],
+    stateValueCache: Cache[DamlStateKey, (DamlStateValue, Fingerprint)],
+    cacheUpdatePolicy: CacheUpdatePolicy)(implicit materializer: Materializer) {
+
+  /**
+    * Pre-executes and then finalizes a submission.
+    */
+  def commit(
+      correlationId: String,
+      submissionEnvelope: Bytes,
+      submittingParticipantId: ParticipantId,
+      ledgerStateAccess: LedgerStateAccess[LogResult])(
+      implicit executionContext: ExecutionContext): Future[SubmissionResult] =
+    // Fidelity level 1: sequential pre-execution. Implemented as: the pre-post-exec pipeline is a single transaction.
+    ledgerStateAccess.inTransaction { ledgerStateOperations =>
+      for {
+        preExecutionOutput <- validator
+          .validate(
+            submissionEnvelope,
+            correlationId,
+            submittingParticipantId,
+            CachingDamlLedgerStateReaderWithFingerprints(
+              stateValueCache,
+              cacheUpdatePolicy,
+              new LedgerStateReaderWithFingerprintsFromValues(
+                ledgerStateOperations,
+                valueToFingerprint),
+              keySerializationStrategy,
+            )
+          )
+        submissionResult <- retry {
+          case PostExecutionFinalizerWithFingerprintsFromValues.Conflict => true
+        } { (_, _) =>
+          postExecutionFinalizer.conflictDetectAndFinalize(
+            now,
+            preExecutionOutput,
+            ledgerStateOperations)
+        }.transform {
+          case Failure(PostExecutionFinalizerWithFingerprintsFromValues.Conflict) =>
+            Success(SubmissionResult.Acknowledged) // But it will simply be dropped.
+          case result => result
+        }
+      } yield submissionResult
+    }
+
+  // TODO consider adding to [[RetryStrategy]] a pre-built exponential backoff strategy
```

Comment:
Please add that to the backlog and remove this TODO.

Reply:
Done.